File: modules/keeper-git/android/src/main/java/com/clikethis123/keeper/KeeperGitBridgeModule.kt
package com.clikethis123.keeper

import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import org.json.JSONArray
import org.json.JSONObject
import org.json.JSONTokener

class KeeperGitBridgeModule : Module() {
  companion object {
    init {
      System.loadLibrary("git_core")
    }
  }

  private external fun clone_git(url: String, path: String): Int
  private external fun git_last_error_message(): String?
  private external fun git_fetch(repoPath: String): Int
  private external fun git_checkout_ex(
    repoPath: String,
    reference: String,
    force: Int,
    noUpdateHead: Int
  ): Int
  private external fun git_current_branch_json(repoPath: String): String?
  private external fun git_list_branches_json(repoPath: String, remote: String?): String?
  private external fun git_merge_json(repoPath: String, optionsJson: String): Int
  private external fun git_commit(repoPath: String, message: String): Int
  private external fun git_push(repoPath: String): Int
  private external fun git_status_json(repoPath: String): String?
  private external fun git_head_oid_json(repoPath: String): String?
  private external fun git_changed_markdown_paths_json(
    repoPath: String,
    fromOid: String,
    toOid: String
  ): String?

  override fun definition() = ModuleDefinition {
    Name("KeeperGitBridge")

    AsyncFunction("clone") { url: String, path: String, promise: Promise ->
      settleCode("clone", clone_git(url, path), promise)
    }

    AsyncFunction("fetch") { repoPath: String, promise: Promise ->
      settleCode("fetch", git_fetch(repoPath), promise)
    }

    AsyncFunction("checkout") { repoPath: String, reference: String, options: Map<String, Any?>?, promise: Promise ->
      val force = if (options?.get("force") == true) 1 else 0
      val noUpdateHead = if (options?.get("noUpdateHead") == true) 1 else 0
      settleCode(
        "checkout",
        git_checkout_ex(repoPath, reference, force, noUpdateHead),
        promise
      )
    }

    AsyncFunction("currentBranch") { repoPath: String, promise: Promise ->
      settlePayload("current_branch", git_current_branch_json(repoPath), promise)
    }

    AsyncFunction("listBranches") { repoPath: String, remote: String?, promise: Promise ->
      settlePayload("list_branches", git_list_branches_json(repoPath, remote), promise)
    }

    AsyncFunction("merge") { repoPath: String, options: Map<String, Any?>, promise: Promise ->
      val ours = options["ours"] as? String
      val theirs = options["theirs"] as? String
      if (ours == null || theirs == null) {
        promise.reject("E_GIT_MERGE", "Rust git merge requires 'ours' and 'theirs'", null)
        return@AsyncFunction
      }

      val payload = JSONObject().apply {
        put("ours", ours)
        put("theirs", theirs)
        if (options.containsKey("fastForwardOnly")) {
          put("fast_forward_only", options["fastForwardOnly"])
        }
        (options["message"] as? String)?.let { put("message", it) }
        val author = options["author"] as? Map<*, *>
        if (author != null) {
          put(
            "author",
            JSONObject().apply {
              put("name", author["name"])
              put("email", author["email"])
            }
          )
        }
      }

      settleCode("merge", git_merge_json(repoPath, payload.toString()), promise)
    }

    AsyncFunction("commit") { repoPath: String, message: String, promise: Promise ->
      settleCode("commit", git_commit(repoPath, message), promise)
    }

    AsyncFunction("push") { repoPath: String, promise: Promise ->
      settleCode("push", git_push(repoPath), promise)
    }

    AsyncFunction("status") { repoPath: String, promise: Promise ->
      settlePayload("status", git_status_json(repoPath), promise)
    }

    AsyncFunction("resolveHeadOid") { repoPath: String, promise: Promise ->
      settlePayload("head_oid", git_head_oid_json(repoPath), promise)
    }

    AsyncFunction("changedMarkdownPaths") { repoPath: String, fromOid: String, toOid: String, promise: Promise ->
      settlePayload(
        "changed_markdown_paths",
        git_changed_markdown_paths_json(repoPath, fromOid, toOid),
        promise
      )
    }
  }

  private fun settleCode(op: String, code: Int, promise: Promise) {
    if (code == 0) {
      promise.resolve(null)
      return
    }

    val detail = consumeLastRustError()?.let { ": $it" } ?: ""
    promise.reject(
      "E_GIT_${op.uppercase()}",
      "Rust git $op failed with code=$code$detail",
      null
    )
  }

  private fun settlePayload(op: String, payload: String?, promise: Promise) {
    if (payload == null) {
      promise.reject("E_GIT_${op.uppercase()}", "Rust git $op failed", null)
      return
    }
    promise.resolve(parseJsonValue(payload))
  }

  private fun consumeLastRustError(): String? {
    val raw = git_last_error_message() ?: return null
    return try {
      val parsed = JSONObject("{\"value\":$raw}").opt("value")
      when (parsed) {
        null -> null
        JSONObject.NULL -> null
        else -> parsed.toString().takeIf { it.isNotBlank() }
      }
    } catch (_: Exception) {
      raw.takeIf { it.isNotBlank() }
    }
  }

  private fun parseJsonValue(payload: String): Any? {
    return convertJsonValue(JSONTokener(payload).nextValue())
  }

  private fun convertJsonValue(value: Any?): Any? {
    return when (value) {
      null, JSONObject.NULL -> null
      is JSONObject -> buildMap {
        val keys = value.keys()
        while (keys.hasNext()) {
          val key = keys.next()
          put(key, convertJsonValue(value.opt(key)))
        }
      }
      is JSONArray -> buildList {
        for (index in 0 until value.length()) {
          add(convertJsonValue(value.opt(index)))
        }
      }
      else -> value
    }
  }
}

File: modules/keeper-git/ios/KeeperGitBridgeModule.swift
import ExpoModulesCore
import Foundation

public final class KeeperGitBridgeModule: Module {
  @_silgen_name("clone_git")
  private func clone_git(_ url: UnsafePointer<CChar>, _ path: UnsafePointer<CChar>) -> Int32

  @_silgen_name("git_fetch")
  private func git_fetch(_ repoPath: UnsafePointer<CChar>) -> Int32

  @_silgen_name("git_checkout_ex")
  private func git_checkout_ex(
    _ repoPath: UnsafePointer<CChar>,
    _ reference: UnsafePointer<CChar>,
    _ force: Int32,
    _ noUpdateHead: Int32
  ) -> Int32

  @_silgen_name("git_current_branch_json")
  private func git_current_branch_json(_ repoPath: UnsafePointer<CChar>) -> UnsafeMutablePointer<CChar>?

  @_silgen_name("git_list_branches_json")
  private func git_list_branches_json(
    _ repoPath: UnsafePointer<CChar>,
    _ remote: UnsafePointer<CChar>?
  ) -> UnsafeMutablePointer<CChar>?

  @_silgen_name("git_merge_json")
  private func git_merge_json(
    _ repoPath: UnsafePointer<CChar>,
    _ optionsJson: UnsafePointer<CChar>
  ) -> Int32

  @_silgen_name("git_commit")
  private func git_commit(_ repoPath: UnsafePointer<CChar>, _ message: UnsafePointer<CChar>) -> Int32

  @_silgen_name("git_push")
  private func git_push(_ repoPath: UnsafePointer<CChar>) -> Int32

  @_silgen_name("git_status_json")
  private func git_status_json(_ repoPath: UnsafePointer<CChar>) -> UnsafeMutablePointer<CChar>?

  @_silgen_name("git_head_oid_json")
  private func git_head_oid_json(_ repoPath: UnsafePointer<CChar>) -> UnsafeMutablePointer<CChar>?


  @_silgen_name("git_changed_markdown_paths_json")
  private func git_changed_markdown_paths_json(
    _ repoPath: UnsafePointer<CChar>,
    _ fromOid: UnsafePointer<CChar>,
    _ toOid: UnsafePointer<CChar>
  ) -> UnsafeMutablePointer<CChar>?

  @_silgen_name("git_changed_paths_json")
  private func git_changed_paths_json(
    _ repoPath: UnsafePointer<CChar>,
    _ fromOid: UnsafePointer<CChar>,
    _ toOid: UnsafePointer<CChar>
  ) -> UnsafeMutablePointer<CChar>?

  @_silgen_name("git_last_error_message")
  private func git_last_error_message() -> UnsafeMutablePointer<CChar>?

  @_silgen_name("git_string_free")
  private func git_string_free(_ ptr: UnsafeMutablePointer<CChar>?)

  public func definition() -> ModuleDefinition {
    Name("KeeperGitBridge")

    AsyncFunction("clone") { (url: String, path: String, promise: Promise) in
      let code = url.withCString { urlPtr in
        path.withCString { pathPtr in
          clone_git(urlPtr, pathPtr)
        }
      }
      settle(code: code, op: "clone", promise: promise)
    }

    AsyncFunction("fetch") { (repoPath: String, promise: Promise) in
      let code = repoPath.withCString { git_fetch($0) }
      settle(code: code, op: "fetch", promise: promise)
    }

    AsyncFunction("checkout") { (repoPath: String, reference: String, options: [String: Any]?, promise: Promise) in
      let force = (options?["force"] as? Bool) == true ? Int32(1) : Int32(0)
      let noUpdateHead = (options?["noUpdateHead"] as? Bool) == true ? Int32(1) : Int32(0)
      let code = repoPath.withCString { repoPtr in
        reference.withCString { refPtr in
          git_checkout_ex(repoPtr, refPtr, force, noUpdateHead)
        }
      }
      settle(code: code, op: "checkout", promise: promise)
    }

    AsyncFunction("currentBranch") { (repoPath: String, promise: Promise) in
      let payloadPtr = repoPath.withCString { git_current_branch_json($0) }
      decodeJsonPayload(payloadPtr: payloadPtr, op: "current_branch", promise: promise)
    }

    AsyncFunction("listBranches") { (repoPath: String, remote: String?, promise: Promise) in
      let payloadPtr: UnsafeMutablePointer<CChar>? = repoPath.withCString { repoPtr in
        if let remote = remote {
          return remote.withCString { remotePtr in
            git_list_branches_json(repoPtr, remotePtr)
          }
        }
        return git_list_branches_json(repoPtr, nil)
      }
      decodeJsonPayload(payloadPtr: payloadPtr, op: "list_branches", promise: promise)
    }

    AsyncFunction("merge") { (repoPath: String, options: [String: Any], promise: Promise) in
      guard let ours = options["ours"] as? String,
            let theirs = options["theirs"] as? String else {
        promise.reject("E_GIT_MERGE", "Rust git merge requires 'ours' and 'theirs'")
        return
      }

      var payload: [String: Any] = [
        "ours": ours,
        "theirs": theirs
      ]
      if let fastForwardOnly = options["fastForwardOnly"] as? Bool {
        payload["fast_forward_only"] = fastForwardOnly
      }
      if let message = options["message"] as? String {
        payload["message"] = message
      }
      if let author = options["author"] as? [String: Any] {
        payload["author"] = [
          "name": author["name"] as Any,
          "email": author["email"] as Any
        ]
      }

      guard JSONSerialization.isValidJSONObject(payload),
            let data = try? JSONSerialization.data(withJSONObject: payload),
            let json = String(data: data, encoding: .utf8) else {
        promise.reject("E_GIT_MERGE", "Rust git merge received invalid options payload")
        return
      }

      let code = repoPath.withCString { repoPtr in
        json.withCString { jsonPtr in
          git_merge_json(repoPtr, jsonPtr)
        }
      }
      settle(code: code, op: "merge", promise: promise)
    }

    AsyncFunction("commit") { (repoPath: String, message: String, promise: Promise) in
      let code = repoPath.withCString { repoPtr in
        message.withCString { msgPtr in
          git_commit(repoPtr, msgPtr)
        }
      }
      settle(code: code, op: "commit", promise: promise)
    }

    AsyncFunction("push") { (repoPath: String, promise: Promise) in
      let code = repoPath.withCString { git_push($0) }
      settle(code: code, op: "push", promise: promise)
    }

    AsyncFunction("status") { (repoPath: String, promise: Promise) in
      let payloadPtr = repoPath.withCString { git_status_json($0) }
      decodeJsonPayload(payloadPtr: payloadPtr, op: "status", promise: promise)
    }

    AsyncFunction("resolveHeadOid") { (repoPath: String, promise: Promise) in
      let payloadPtr = repoPath.withCString { git_head_oid_json($0) }
      decodeJsonPayload(payloadPtr: payloadPtr, op: "head_oid", promise: promise)
    }


    AsyncFunction("changedMarkdownPaths") { (repoPath: String, fromOid: String, toOid: String, promise: Promise) in
      let payloadPtr = repoPath.withCString { repoPtr in
        fromOid.withCString { fromOidPtr in
          toOid.withCString { toOidPtr in
            git_changed_markdown_paths_json(repoPtr, fromOidPtr, toOidPtr)
          }
        }
      }
      decodeJsonPayload(payloadPtr: payloadPtr, op: "changed_markdown_paths", promise: promise)
    }

    AsyncFunction("changedPaths") { (repoPath: String, fromOid: String, toOid: String, promise: Promise) in
      let payloadPtr = repoPath.withCString { repoPtr in
        fromOid.withCString { fromOidPtr in
          toOid.withCString { toOidPtr in
            git_changed_paths_json(repoPtr, fromOidPtr, toOidPtr)
          }
        }
      }
      decodeJsonPayload(payloadPtr: payloadPtr, op: "changed_paths", promise: promise)
    }
  }

  private func decodeJsonPayload(payloadPtr: UnsafeMutablePointer<CChar>?, op: String, promise: Promise) {
    guard let payloadPtr else {
      promise.reject("E_GIT_\(op.uppercased())", "Rust git \(op) failed")
      return
    }

    let payload = String(cString: payloadPtr)
    git_string_free(payloadPtr)

    guard let data = payload.data(using: .utf8),
          let decoded = try? JSONSerialization.jsonObject(with: data) else {
      promise.reject("E_GIT_\(op.uppercased())", "Rust git \(op) returned invalid JSON")
      return
    }

    promise.resolve(decoded)
  }

  private func settle(code: Int32, op: String, promise: Promise) {
    if code == 0 {
      promise.resolve(nil)
    } else {
      let detail = consumeLastRustError().map { ": \($0)" } ?? ""
      promise.reject("E_GIT_\(op.uppercased())", "Rust git \(op) failed with code=\(code)\(detail)")
    }
  }

  private func consumeLastRustError() -> String? {
    guard let errorPtr = git_last_error_message() else {
      return nil
    }
    let message = String(cString: errorPtr)
    git_string_free(errorPtr)
    return message.isEmpty ? nil : message
  }
}
